web/client-api/src/test/java/io/deephaven/web/client/api/AbstractAsyncGwtTestCase.java

```diff
@@ -0,0 +1,391 @@
+            String... events) {
+        RemoverFn[] undos = new RemoverFn[events.length];
+        for (int i = 0; i < events.length; i++) {
+            final String ev = events[i];
+            undos[i] = handling.addEventListener(ev, e -> {
+                log("Did not expect", ev, "but fired event", e);
+                report("Expected " + ev + " to not be called; detail: " + (e.detail));
+            });
+        }
+        return () -> {
+            for (RemoverFn undo : undos) {
+                undo.remove();
+            }
+
+        };
+    }
+
+    static <T> IThenable.ThenOnFulfilledCallbackFn<T, T> run(JsRunnable allow) {
+        return t -> {
+            allow.run();
+            return Promise.resolve(t);
+        };
+    }
+
+    static <T> Promise<T> expectFailure(Promise<?> state, T value) {
+        return state.then(val -> Promise.reject("Failed"),
+                error -> Promise.resolve(value));
+    }
+
+    /**
+     * Imports the webpack content, including protobuf types. Does not connect to the server.
+     */
+    protected Promise<Void> setupDhInternal() {
+        delayTestFinish(500);
+        return importDhInternal();
+    }
+
+    /**
+     * Connects and authenticates to the configured server and runs the specified scripts.
+     */
+    protected Promise<IdeSession> connect(TableSourceBuilder tables) {
+        // start by delaying test finish by .5s so we fail fast in cases where we aren't set up right
+        delayTestFinish(500);
+        return importDhInternal().then(module -> {
+            CoreClient coreClient = new CoreClient(localServer, null);
+            return coreClient.login(JsPropertyMap.of("type", CoreClient.LOGIN_TYPE_ANONYMOUS))
+                    .then(ignore -> coreClient.getAsIdeConnection())
+                    .then(ide -> {
+                        delayTestFinish(501);
+                        return ide.getConsoleTypes().then(consoleTypes -> {
+                            delayTestFinish(502);
+                            CancellablePromise<IdeSession> ideSession = ide.startSession(consoleTypes.getAt(0));
+                            return ideSession.then(session -> {
+
+                                if (consoleTypes.includes("python")) {
+                                    return runAllScriptsInOrder(ideSession, session, tables.pythonScripts);
+                                }
+                                throw new IllegalStateException("Unsupported script type " + consoleTypes);
+                            });
+                        });
+                    });
+        });
+    }
+
+    private Promise<IdeSession> runAllScriptsInOrder(CancellablePromise<IdeSession> ideSession, IdeSession session,
+            List<String> code) {
+        Promise<IdeSession> result = ideSession;
+        for (int i = 0; i < code.size(); i++) {
+            final int index = i;
+            result = result.then(ignore -> {
+                delayTestFinish(4000 + index);
```

Comment:
Constant? Unsure why 4000 was selected.

Reply:
It seems that running python on a brand new server in docker can sometimes be a little slow out of the gate, so I have a "large" timeout here. 

It could make sense to change the strategy entirely and just have a per-test timeout instead, and set it to a few minutes - but it is nice to fail within a few seconds of the failing operation (as we don't have event handlers for all failures, we just let timeouts expire).

Reply:
Yea that's all fine and good, just as a constant you can have those details as a comment so it's clear why you want a timeout of a 4s instead of say 40s or 400ms

Reply:
All of these timeouts are "worst case, sometimes the browser or dh api lags a bit, here's a reasonably sized timeout that likely can't be exceeded in reasonable circumstances". 40ms definitely isn't good enough (browser GC could kill that alone), 400ms probably isn't enough either. 

We _could_ just say "every test should take less than 5 minutes" and not worry about each operation, but historically this has made it easier to figure out which specific operation is stuck/failing.

Reply:
Right - that all makes sense, now add that "why" information as a comment on that timeout.